runtime/etc/mli/mli_common_code.c

```diff
@@ -0,0 +1,196 @@
+#define CHPL_RUNTIME_ETC_MLI_MLI_COMMON_CODE_C_
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdarg.h>
+#include <assert.h>
+#include <zmq.h>
+
+//
+// Hopeful workaround to prevent Travis from complaining. We may just have to
+// force the client to also use the Chapel runtime as well, for better or
+// for worse.
+//
+#ifdef CHPL_MLI_IS_CLIENT_
+# define mli_malloc(bytes) chpl_malloc(bytes)
+# define mli_free(ptr) chpl_free(ptr)
+#elif defined(CHPL_MLI_IS_SERVER_)
+# define mli_malloc(bytes) malloc(bytes)
+# define mli_free(ptr) free(ptr)
+# else
+# error The mli_malloc/free macros were defined in a inappropriate place.
+# endif
+
+
+// Error codes must all be less than zero unless we change the protocol!
+enum chpl_mli_errors {
+
+  CHPL_MLI_ERROR_NONE       = +0,
+  CHPL_MLI_ERROR_SHUTDOWN   = -1,
+  CHPL_MLI_ERROR_UNKNOWN    = -2,
+  CHPL_MLI_ERROR_NOFUNC     = -3,
+  CHPL_MLI_ERROR_SOCKET     = -4,
+  CHPL_MLI_ERROR_EXCEPT     = -5
+
+};
+
+const char* chpl_mli_errstr(enum chpl_mli_errors e) {
+  static const char* mli_errors_[] = {
+    "NONE",
+    "SHUTDOWN",
+    "UNKNOWN",
+    "NOFUNC",
+    "SOCKET",
+    "EXCEPT"
+  };
+
+  if (e > CHPL_MLI_ERROR_NONE || e < CHPL_MLI_ERROR_EXCEPT) {
+    return "INVALID_ERROR_CODE";
+  }
+
+  return mli_errors_[e];
+}
+
+//
+// Both the client and server will be using this to perform RPC.
+//
+struct chpl_mli_context {
+
+  void* context;
+  void* main;
+  void* arg;
+  void* res;
+
+};
+
+static
+char* chpl_mli_concat(size_t count, ...) {
+  char* result = NULL;
+  char* bufptr = NULL;
+  va_list argp;
+  size_t i = 0;
+  size_t length = 0;
+
+  va_start(argp, count);
+
+  for (i = 0; i < count; i++) {
+    const char* chunk = va_arg(argp, const char*);
+    length += strlen(chunk);
+  }
+
+  result = mli_malloc(length + 1);
+  if (result == NULL) { return NULL; }
+
+  va_start(argp, count);
+  bufptr = result;
+
+  for (i = 0; i < count; i++) {
+    const char* chunk = va_arg(argp, const char*);
+    strcpy(bufptr, chunk);
+    bufptr += strlen(chunk);
+  }
+
+  result[length] = 0;
+  va_end(argp);
+
+  return result;  
+}
+
+//
+// In this interface only the server should bind!
+//
+static
+void chpl_mli_bind(void* socket, const char* ip, const char* port) {
+  const char* real_ip = ip;
+  char* buffer = NULL;
+  char* bufptr = NULL;
+  size_t len = 0;
+  int err = 0;
+
+  //
+  // The ZMQ API is non-orthagonal in this way, so in our interface we opt to
+  // substitute "localhost" for "*" (indicating any device) automatically.
+  // Note that IP addresses are not otherwise sanitized (ZMQ will do that for
+  // us, but we don't return errors yet).
+  //
+  if (!strcmp(ip, "localhost")) {
+    real_ip = "*";
+  }
+
+  buffer = chpl_mli_concat(4, "tcp://", real_ip, ":", port);
+
+  // TODO: Handle malloc error here?
+  if (buffer == NULL) {
+    assert("Internal malloc failed, should not reach (yet)." == 0);
```

Comment:
This line seems a little wonky to me.  I don't understand what the last half of it means, and I suspect there are better ways to print error messages to the user

Reply:
So that's just a technique to print a message to the screen with an assert (it always evaluates to true). It's probably not the best thing to do, as we'd probably want to shut down the server as well, but when I was working on the prototype my logic was along the lines of: "well, if a memory allocation error occurs, we're probably done for, so just crash".

Reply:
The bigger question here is how malloc failures will be handled.

Reply:
So the problem in this particular scenario is that if a malloc fails (which it probably shouldn't, but who knows), then that particular socket won't be able to bind.

Since this is during server setup, without binding the sockets, the server may not be able to communicate that failure back to the client.

There are multiple solutions to this problem, I'm not sure which one is best.

Reply:
That also depends on how we spawn the server, though.  In the strategy I took, I believe there was a way to poll the subprocess spawned to see if it was still active - it's quite possible there will be a way to do that in the implementation we follow there (and it should be listed in the things to think about for that task - e.g. I don't think you have to think through the full ramifications now)